pkg/pillar/cmd/zedagent/handleProfile.go

```diff
@@ -0,0 +1,259 @@
+}
+
+// Run a periodic fetch of the currentProfile
+func localProfileTimerTask(handleChannel chan interface{}, getconfigCtx *getconfigContext) {
+
+	ctx := getconfigCtx.zedagentCtx
+	iteration := 0
+	localProfileServer := getconfigCtx.localProfileServer
+	if localProfileServer != "" {
+		localProfileURL, err := getLocalProfileURL(localProfileServer)
+		if err != nil {
+			log.Errorf("localProfileTimerTask: getLocalProfileURL: %s", err)
+		} else {
+			if err := getLocalProfileConfig(localProfileURL, iteration, getconfigCtx); err != nil {
+				log.Errorf("localProfileTimerTask: %s", err)
+			}
+		}
+		publishZedAgentStatus(getconfigCtx)
+	} else {
+		cleanSavedProtoMessage(savedLocalProfileFile)
+	}
+
+	// use ConfigInterval as localProfileInterval
+	localProfileInterval := ctx.globalConfig.GlobalValueInt(types.ConfigInterval)
+	interval := time.Duration(localProfileInterval) * time.Second
+	max := float64(interval)
+	min := max * 0.3
+	ticker := flextimer.NewRangeTicker(time.Duration(min),
+		time.Duration(max))
+	// Return handle to caller
+	handleChannel <- ticker
+
+	wdName := agentName + "currentProfile"
+
+	// Run a periodic timer so we always update StillRunning
+	stillRunning := time.NewTicker(25 * time.Second)
+	ctx.ps.StillRunning(wdName, warningTime, errorTime)
+	ctx.ps.RegisterFileWatchdog(wdName)
+
+	for {
+		select {
+		case <-ticker.C:
+			localProfileServer = getconfigCtx.localProfileServer
+			if localProfileServer != "" {
+				start := time.Now()
+				iteration++
+				localProfileURL, err := getLocalProfileURL(localProfileServer)
+				if err != nil {
+					log.Errorf("getLocalProfileURL: %s", err)
+				} else {
+					if err := getLocalProfileConfig(localProfileURL, iteration, getconfigCtx); err != nil {
+						log.Errorf("localProfileTimerTask getLocalProfileConfig: %s", err)
+					} else {
+						publishZedAgentStatus(getconfigCtx)
+					}
+				}
+				ctx.ps.CheckMaxTimeTopic(wdName, "getLocalProfileConfig", start,
+					warningTime, errorTime)
+			}
+
+		case <-stillRunning.C:
+		}
+		ctx.ps.StillRunning(wdName, warningTime, errorTime)
+	}
+}
+
+func validateAndSetLocalProfile(localProfileBytes []byte, getconfigCtx *getconfigContext) error {
+	var localProfile = &profile.LocalProfile{}
+	err := proto.Unmarshal(localProfileBytes, localProfile)
+	if err != nil {
+		return fmt.Errorf("validateAndSetLocalProfile: Unmarshalling failed: %v", err)
+	}
+	if localProfile.GetServerToken() != getconfigCtx.profileServerToken {
+		// send something to ledmanager ??
+		return fmt.Errorf("validateAndSetLocalProfile: missamtch ServerToken for loacl profile server")
+	}
+	if getconfigCtx.currentProfile != localProfile.GetLocalProfile() {
+		log.Noticef("validateAndSetLocalProfile: profile changed from %s to %s",
+			getconfigCtx.currentProfile, localProfile.GetLocalProfile())
+		getconfigCtx.currentProfile = localProfile.GetLocalProfile()
+		writeReceivedProtoMessageLocalProfile(localProfileBytes)
+		publishZedAgentStatus(getconfigCtx)
```

Comment:
This should move to the timer loop to make make it visible and avoid side effects. validateXXX() should just return error and not have side effects.

Reply:
Means function name should be validateLocalProfile()